Snakefile

```diff
@@ -524,58 +676,170 @@ if augment_corpus:
-        teacher_models=expand(f"{final_teacher_dir}{{ens}}/{best_model}",ens=ensemble)
-    output: f'{translated}/corpus/file.{{part}}.nbest'
-    params: args=get_args('decoding-teacher')
-    shell: '''bash pipeline/translate/translate-nbest.sh \
-                "{input.file}" "{input.vocab}" {input.teacher_models} {params.args} >> {log} 2>&1'''
+# Workflow without fine-tuning teacher models to parallel corpora
+# Forward-translate with general teacher
+if not fine_tune_to_corpus:
+    checkpoint split_corpus:
+        message: "Splitting the corpus to translate"
+        log: f"{log_dir}/split_corpus.log"
+        conda: "envs/base.yml"
+        threads: 1
+        input: corpus_src=clean_corpus_src,corpus_trg=clean_corpus_trg
+        output: directory(f"{translated}/corpus")
+        shell: '''bash pipeline/translate/split-corpus.sh \
+                    {input.corpus_src} {input.corpus_trg} {output} {split_length} >> {log} 2>&1'''
+
+    rule translate_corpus:
+        message: "Translating corpus with teacher"
+        log: f"{log_dir}/translate_corpus/{{part}}.log"
+        conda: "envs/base.yml"
+        threads: gpus_num*2
+        resources: gpu=gpus_num
+        input:
+            ancient(decoder),
+            file=f'{translated}/corpus/file.{{part}}',
+            vocab=vocab_path,
+            teacher_models=expand(f"{final_teacher_dir}{{ens}}/{best_model}",ens=ensemble)
+        output: f'{translated}/corpus/file.{{part}}.nbest'
+        params: args=get_args('decoding-teacher')
+        shell: '''bash pipeline/translate/translate-nbest.sh \
+                    "{input.file}" "{input.vocab}" {input.teacher_models} {params.args} >> {log} 2>&1'''
+
+    rule extract_best:
+        message: "Extracting best translations for the corpus"
+        log: f"{log_dir}/extract_best/{{part}}.log"
+        conda: "envs/base.yml"
+        threads: 1
+        #group 'translate_corpus'
+        input: nbest=f"{translated}/corpus/file.{{part}}.nbest", ref=f"{translated}/corpus/file.{{part}}.ref"
+        output: f"{translated}/corpus/file.{{part}}.nbest.out"
+        shell: 'python pipeline/translate/bestbleu.py -i {input.nbest} -r {input.ref} -m bleu -o {output} >> {log} 2>&1'
+
+    rule collect_corpus:
+        message: "Collecting translated corpus"
+        log: f"{log_dir}/collect_corpus.log"
+        conda: "envs/base.yml"
+        threads: 4
+        #group 'translate_corpus'
+        input:
+            lambda wildcards: expand(f"{translated}/corpus/file.{{part}}.nbest.out",
+                part=find_parts(wildcards, checkpoints.split_corpus))
+        output: f'{translated}/corpus.{trg}.gz'
+        params: src_corpus=clean_corpus_src
+        shell: 'bash pipeline/translate/collect.sh {translated}/corpus {output} {params.src_corpus} >> {log} 2>&1'
+
+# Workflow with fine-tuning teachers to multiple parallel corpora
+# Translate with multiple domain teachers
+if fine_tune_to_corpus:
+    checkpoint split_corpus_domains:
```

Comment:
I'm concerned about workflow complexity and duplication of its big chunk. These 4 rules appear to be the same for both scenarios. The only difference is that the second ones use `{{dataset}}` wildcard. I think if we leave just them and use variables for inputs/outputs Snakmake will be able to infer that it has to use those rules for both scenarios (basically `corpus` will be just a special case of `{{dataset}}`).

Reply:
Totally agree that this doesn't look good. This is the best I've come up with so far, but I'll look into it again and see if I can do this with better wildcard usage and less code